### PR TITLE
Fix embedding example, simplify Makefile, and test it

### PR DESCRIPTION
**Pull request by ihnorton**
- get rid of makefile cruft
- build in-place instead of putting in usr/bin
- use julia-config for flags, as documented in manual

(related: #18421)

### CODE REVIEW

File: examples/Makefile [hunk 1]
```diff
@@ -1,41 +1,25 @@
-JULIAHOME = $(abspath ..)
-include $(JULIAHOME)/Make.inc
+JULIA = $(abspath ../usr/bin/julia)
```
Comment (tkelman) [inline examples/Makefile]: this isn't right for an out-of-tree build

File: examples/Makefile [hunk 2]
```diff
@@ -1,41 +1,25 @@
...
+JL_SHARE = $(shell $(JULIA) -e 'print(joinpath(JULIA_HOME,Base.DATAROOTDIR,"julia"))')
```
Comment (tkelman) [inline examples/Makefile]: may require spawn

File: examples/Makefile [hunk 3]
```diff
@@ -1,41 +1,25 @@
...
 
-DEBUGFLAGS += $(FLAGS)
-SHIPFLAGS += $(FLAGS)
-JLDFLAGS += $(LDFLAGS) $(NO_WHOLE_ARCHIVE) $(call exec,$(LLVM_CONFIG) --ldflags) $(OSLIBS) $(RPATH)
-
-ifeq ($(USE_SYSTEM_LIBM),0)
-ifneq ($(UNTRUSTED_SYSTEM_LIBM),0)
-JLDFLAGS += $(WHOLE_ARCHIVE) $(build_libdir)/libopenlibm.a $(NO_WHOLE_ARCHIVE)
-endif
+ifeq ($(OS),Windows_NT)
```
Comment (tkelman) [inline examples/Makefile]: not correct for cross-compiling

File: src/julia.h [hunk 4]
```diff
@@ -1313,6 +1313,7 @@ JL_DLLEXPORT void julia_init(JL_IMAGE_SEARCH rel);
 JL_DLLEXPORT void jl_init(const char *julia_home_dir);
 JL_DLLEXPORT void jl_init_with_image(const char *julia_home_dir,
                                      const char *image_relative_path);
+JL_DLLEXPORT const char* jl_get_default_sysimg_path();
```
Comment (tkelman) [inline src/julia.h]: if exported, needs JL_DLLEXPORT on both prototype and implementation

**Comment by ihnorton** [comment]
Updated, thanks. I'm going for non-reliance on `Make.inc`, so that this can serve as a usable template. However, added checks for spawn and BUILDROOT independently (to handle cross-compile and out-of-tree). Hopefully sufficient, though I'm not set up to test those right now.

I'll note that this example isn't built at all right now with out-of-tree.

**Comment by ihnorton** [comment]
Win64 and OSX test failures seem unrelated.

File: examples/Makefile [hunk 5]
```diff
@@ -1,41 +0,0 @@
...
-FLAGS = -Wall -Wno-strict-aliasing -fno-omit-frame-pointer \
-	-I$(JULIAHOME)/src -I$(JULIAHOME)/src/support -I$(build_includedir) $(CFLAGS)
-
-DEBUGFLAGS += $(FLAGS)
-SHIPFLAGS += $(FLAGS)
-JLDFLAGS += $(LDFLAGS) $(NO_WHOLE_ARCHIVE) $(call exec,$(LLVM_CONFIG) --ldflags) $(OSLIBS) $(RPATH)
-
-ifeq ($(USE_SYSTEM_LIBM),0)
-ifneq ($(UNTRUSTED_SYSTEM_LIBM),0)
-JLDFLAGS += $(WHOLE_ARCHIVE) $(build_libdir)/libopenlibm.a $(NO_WHOLE_ARCHIVE)
```
Comment (tkelman) [inline examples/Makefile]: does julia-config handle this and the llvm-config output?

File: test/examples.jl [hunk 6]
```diff
@@ -89,3 +89,10 @@ if !zmq_found
 end
 
 include(joinpath(dir, "clustermanager/0mq/ZMQCM.jl"))
+
+# test that the embedding example runs without error
+let
+    lines = readlines(pipeline(`$(joinpath(JULIA_HOME,Base.DATAROOTDIR,"doc","julia","examples","embedding","embedding"))`, stderr=DevNull))
```
Comment (tkelman) [inline test/examples.jl]: wrap long line

File: examples/embedding/Makefile [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+debug:   embedding-debug$(EXE)
+
+embedding$(EXE): $(SRCDIR)/embedding.c
+	$(CC) $^ -o $@ $(CFLAGS)
+
+embedding-debug$(EXE): $(SRCDIR)/embedding.c
+	$(CC) $^ -o $@ $(CFLAGS) $(DEBUGFLAGS)
+
+clean:
+	rm -f *.o *.do
```
Comment (tkelman) [inline examples/embedding/Makefile]: doesn't look like this will create any object files any more

**Comment by tkelman** [comment]
we shouldn't install binaries in the same place as documentation and source code. installing and redistributing the compiled embedding results isn't the most useful test of this, see my comments on https://github.com/JuliaLang/julia/pull/18688

File: examples/Makefile (hunk 5 above)
Comment (ihnorton) [inline examples/Makefile]: libm yes, llvm-config no. Doesn't libjulia link llvm already?

Comment (tkelman) [inline examples/Makefile]: yes but presumably the llvm-config --ldflags call was needed for something

Comment (ihnorton) [inline examples/Makefile]: That was added in 072a861828378 when the file was created. My guess is copy-paste or prior/local build issue. If it breaks the example we can put it back.

**Comment by ihnorton** [comment]
> installing and redistributing the compiled embedding results isn't the most useful test of this, see my comments on #18688

Ok:
- I put the binary in libexec in the latest update, because I agree with Yichao (and your thumbs-up) in that thread that this could be a useful test of relocatability, which I will have a follow-up PR to make possible.
- I made the test optional, as you requested in that thread, and added to appveyor/travis scripts.

File: examples/embedding/Makefile [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  JULIA = $(error "Must pass JULIA=[target julia binary], or set environment variable!")
+else
+  EXE = $(suffix $(abspath $(JULIA)))
+endif
+
+#=============================================================================
+# hooks for julia build system
+ifdef BUILDROOT
+  SRCDIR := $(abspath $(dir $(lastword $(MAKEFILE_LIST))))/
+  BINDIR := $(abspath $(BUILDROOT)/usr/libexec)/
```
Comment (tkelman) [inline examples/embedding/Makefile]: `build_libexecdir`

File: examples/embedding/Makefile [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# This Makefile template requires the following variables to be set
+# in the environment or on the command-line:
+#   JULIA: path to julia[.exe] executable
+#   BIN:   binary build directory
+
+ifndef JULIA
+  $(error "Please set JULIA to the path of the target julia binary!")
+endif
+ifndef BIN
+  $(error "Please set BIN to the path of the build directory!")
```
Comment (tkelman) [inline examples/embedding/Makefile]: could maybe use a bit more clarification what is meant by this

File: examples/embedding/Makefile [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+#=============================================================================
+
+release: $(BIN)/embedding$(EXE)
+debug:   $(BIN)/embedding-debug$(EXE)
+
+$(BIN)/embedding$(EXE): embedding.c
+	$(CC) $^ -o $@ $(CFLAGS)
+
+$(BINDIR)/embedding-debug$(EXE): embedding.c
```
Comment (tkelman) [inline examples/embedding/Makefile]: should this also be BIN ?

File: examples/embedding/Makefile [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+DEBUGFLAGS += -g
+
+#=============================================================================
+
+release: $(BIN)/embedding$(EXE)
+debug:   $(BIN)/embedding-debug$(EXE)
+
+$(BIN)/embedding$(EXE): embedding.c
+	$(CC) $^ -o $@ $(CFLAGS)
```
Comment (tkelman) [inline examples/embedding/Makefile]: might want to also explicitly error if CC isn't set, or default to `cc`

Comment (vtjnash) [inline examples/embedding/Makefile]: `make` already does that

File: contrib/julia-config.jl [hunk 12]
```diff
@@ -70,9 +70,9 @@ function cflags()
     arg2 = replace(includeDir(),"\\","\\\\")
     threading_def = threadingOn() ? "-DJULIA_ENABLE_THREADING=1 " : ""
     if is_unix()
-        return """$(threading_def)-fPIC -DJULIA_INIT_DIR=\\"$arg1\\" -I$arg2"""
+        return """-std=gnu99 $(threading_def)-fPIC -DJULIA_INIT_DIR=\\"$arg1\\" -I$arg2"""
```
Comment (tkelman) [inline contrib/julia-config.jl]: does the runtime need this, or the code in the embedding example?

Comment (ihnorton) [inline contrib/julia-config.jl]: the runtime.

Comment (tkelman) [inline contrib/julia-config.jl]: are we using gnu extensions in our headers, or which compiler versions don't default to c99 mode?

File: examples/embedding/embedding.c [hunk 13]
```diff
@@ -14,7 +14,7 @@ double my_c_sqrt(double x)
 
 int main()
 {
-    jl_init(NULL);
+    jl_init(JULIA_INIT_DIR);
```
Comment (JeffBezanson) [inline examples/embedding/embedding.c]: Does this still work with `NULL` as the argument?

Comment (JeffBezanson) [inline examples/embedding/embedding.c]: Oh, even better, it looks like `jl_init` takes no arguments now?

File: src/jlapi.c [hunk 14]
```diff
@@ -43,13 +43,28 @@ JL_DLLEXPORT void jl_init_with_image(const char *julia_home_dir,
...
     if (image_relative_path != NULL)
         jl_options.image_file = image_relative_path;
+    else
+        jl_options.image_file = jl_get_default_sysimg_path();
     julia_init(JL_IMAGE_JULIA_HOME);
     jl_exception_clear();
 }
 
-JL_DLLEXPORT void jl_init(const char *julia_home_dir)
+JL_DLLEXPORT void jl_init()
```
Comment (JeffBezanson) [inline src/jlapi.c]: Should be `jl_init(void)`.

File: src/julia.h [hunk 15]
```diff
@@ -1309,12 +1309,15 @@ typedef enum {
...
-JL_DLLEXPORT void jl_init(const char *julia_home_dir);
+JL_DLLEXPORT void jl_init();
 JL_DLLEXPORT void jl_init_with_image(const char *julia_home_dir,
                                      const char *image_relative_path);
+JL_DLLEXPORT const char* jl_get_default_sysimg_path();
 JL_DLLEXPORT int jl_is_initialized(void);
 JL_DLLEXPORT void jl_atexit_hook(int status);
 JL_DLLEXPORT void JL_NORETURN jl_exit(int status);
+JL_DLLEXPORT const char* jl_pathname_for_handle(void *handle);
+JL_DLLEXPORT const char* jl_get_default_sysimg_path();
```
Comment (JeffBezanson) [inline src/julia.h]: This is duplicated and needs `(void)`.

**Review by JeffBezanson — CHANGES_REQUESTED** [review]

**Comment by JeffBezanson** [comment]
Bump.

**Comment by JeffBezanson** [comment]
Moved to #21299.

**Pull request closed without merging**